mdformat/_util.py

```diff
@@ -1,21 +1,81 @@
-import re
-from typing import Iterable
+from html.parser import HTMLParser
+from typing import Iterable, List, Optional, Set
 
 from markdown_it import MarkdownIt
 
+import mdformat.plugins
 
-def is_md_equal(md1: str, md2: str, *, ignore_codeclasses: Iterable[str] = ()) -> bool:
+
+def is_md_equal(
+    md1: str,
+    md2: str,
+    *,
+    enabled_extensions: Iterable[str] = (),
+    enabled_codeformatters: Iterable[str] = (),
+) -> bool:
     """Check if two Markdown produce the same HTML.
 
-    Renders HTML from both Markdown strings, strips whitespace and
-    checks equality. Note that this is not a perfect solution, as there
-    can be meaningful whitespace in HTML, e.g. in a <code> block.
+    Renders HTML from both Markdown strings, strip content of tags with
+    specified classes, and checks equality.
     """
-    html1 = MarkdownIt().render(md1)
-    html2 = MarkdownIt().render(md2)
-    html1 = re.sub(r"\s+", "", html1)
-    html2 = re.sub(r"\s+", "", html2)
-    for codeclass in ignore_codeclasses:
-        html1 = re.sub(rf'<codeclass="language-{codeclass}">.*</pre>', "", html1)
-        html2 = re.sub(rf'<codeclass="language-{codeclass}">.*</pre>', "", html2)
+    ignore_classes = [f"language-{lang}" for lang in enabled_codeformatters]
+    mdit = MarkdownIt()
+    for name in enabled_extensions:
+        plugin = mdformat.plugins.PARSER_EXTENSIONS[name]
+        plugin.update_mdit(mdit)
+        ignore_classes.extend(getattr(plugin, "ignore_classes", []))
```

Comment:
This is new plugin API and should be documented if we leave it in. What would this be used for though? There aren't many classes in HTML generated from Markdown.

I've been thinking to make a TOC generator plugin where I would need to ignore everything in the HTML that's in between certain types of HTML comments. A regex based ignore system could work there, or a flag that disables equality check altogether, but not sure if a class based system like this would.

Reply:
why in between HTML comments? what would the ToC HTML structure look like?
and anyway, if we drop `handle_comment` (above), then comments would be ignored anyway?

well you could have something like the amsmath plugin that generates:

```html
<section class="amsmath">
<eqn>
\begin{equation*}
a = 1
\end{equation*}
</eqn>
</section>
```

and so if you wanted to format the TeX in some way, you would specify `ignore_classes = ["amsmath"]`. You could even take it one step further and specify CSS selector like syntax, e.g. `section.amsmath` (this is available in beautifulsoup)

Reply:
I can add it to the `ParserExtensionInterface` though

Reply:
> what would the ToC HTML structure look like?

Something like https://github.com/thlorenz/doctoc#specifying-location-of-toc

So you leave these comments in the Markdown and the tool will generate a TOC in between them.

Reply:
Hmm, I'm thinking maybe just a boolean switch to turn off the Markdown equality check altogether might be the cleanest solution that would work for every possible use case, and with very little effort...?

Reply:
yeh that feels like a bit of an edge case that we can deal with later

Reply:
BTW it feels like there could be some possible integration of the ToC plugin and https://github.com/executablebooks/markdown-it-py/blob/master/markdown_it/extensions/anchors/index.py

Reply:
i.e. potentially moving the anchor plugin code out of markdown-it-py and into the same package as the toc generator plugin

Reply:
Ah so what I'm suggesting already now, before we add the class-based-skip-equality-check API, should we just instead give plugins a boolean flag they can set that completely disables the equality check. The ToC plugin for instance seems to need something more powerful than the class-based system, and I wouldn't want to end up supporting many ways to bypass the equality check.

This way:
- the API is dead simple
- we never have to extend it

The whole equality check is just a safeguard really, intended for the default "pure" formatter mode, that is not supposed to alter AST.

Reply:
I prefer to keep the equality check where possible. I think for most plugins the `ignore_classes` option should be sufficient.
This ToC plugin feels like a bit of special case, in that it is actually adding something new to the document, rather than just formatting what is already there

Reply:
Ok so I thought about this a bit. Current thoughts:

- I will want to add a `CHANGES_AST` (or similar name) switch to the plugin API in any case.
- The equality check is really just a safeguard for the CLI (currently not applied to Python API), not a core part of mdformat. Assuming no bugs, the feature does nothing. I want to avoid overengineering a set of plugin hooks to alter that.
- The formatter in theory should never alter AST/rendered markdown. I feel like any plugin that does so, is equally a special case.

So I'm thinking to reject the class-based plugin hook. I know this could be a bit frustrating, sorry for that. And if so, I encourage to plan new features in an issue before implementation.

Reply:
could the equality check be toggled to only check the equality of the HTML AST structure, rather than its full content, i.e. `def handle_data` would not store anything in this mode.
This would still check that the formatting does not change the overall tag structure of the HTML (which a plugin should not generally do), but would not worry about its content (which mdformat core and plugins literally all change).

e.g. it would just check that both docs have the structure

```html
<div a="1">
  <other b="2">
  </other>
</div>
```

rather than check that they are exactly:

```html
<div a="1">
  a b
  c d
  <other b="2">
  x y z
  </other>
</div>
```

Reply:
> The formatter in theory should never alter AST/rendered markdown

and yeh just to emphasise; in some sense the formatter *always* alters the HTML AST, if you are taking into account the content of each tag, e.g. stripping whitespace in this context is altering the AST.

what the formatter should not do (in theory) is alter the HTML AST structure

Reply:
> could the equality check be toggled to only check the equality of the HTML AST structure, rather than its full content

This would also require an addition to the API, which I'd rather not do. I've thought about possibly even removing the equality check once mdformat proves to be stable and bug free (since it makes the tool render 3 docs instead of just 1). I'd prefer the API to not be littered with stuff related.

> This would still check that the formatting does not change the overall tag structure of the HTML (which a plugin should not generally do), but would not worry about its content (which mdformat core and plugins literally all change).

> stripping whitespace in this context is altering the AST.

I mean CommonMark AST, which AFAIK should completely ignore leading or trailing whitespace. The HTML rendering is just a tool used to check that.
Besides leading/trailing whitespace the changes mdformat does are mostly just adding escape characters and converting characters to their numeric representation (which AFAIK should not change HTML in any way).

> what the formatter should not do (in theory) is alter the HTML AST structure

It's more strict than this. All ASTs aside, what I really want is the rendered HTML to look precisely the same before and after running mdformat (core). That means that the changes mentioned above are fine (when applied to paragraphs), but most other changes to content (e.g. any changes to code blocks) are not. Plugins are free to do what they want of course.

Reply:
> I've thought about possibly even removing the equality check

I don't know, I quite like it. You probably already know, but black has a similar equivalence check
https://github.com/psf/black/blob/172c0a78facd8f938629f216c95507bbfeddfe5e/src/black/__init__.py#L944

> I mean CommonMark AST, which AFAIK should completely ignore leading or trailing whitespace.

I'm not sure if this is correct. I mean there's not really a CommonMark AST anyway, just a specification of what HTML it should produce.

It would just be nice if you didn't have to completely throw away this check, due to the use of one plugin.

Reply:
> I'm not sure if this is correct. I mean there's not really a CommonMark AST anyway, just a specification of what HTML it should produce.

I'm pretty sure every language can be parsed into a syntax tree, even if there's not an established format how it should look like. According to CommonMark spec `The paragraph’s raw content is formed by concatenating the lines and removing initial and final whitespace.`, so a parser will definitely remove leading/trailing whitespace and it won't be in this syntax tree.

> It would just be nice if you didn't have to completely throw away this check, due to the use of one plugin.

With the `CHANGES_AST` flag, you don't have throw it away, unless the plugin makes changes to how the rendered Markdown looks like. That seems like simple and reasonable spec to me.

I think at this point we should just agree to disagree? :smile: :peace_symbol: